### DISCUSSION TITLE
Need Guidance : how to create a reusable expression and pass JSON into the engine for quick evaluation

### DISCUSSION THREAD

**Discussion by rajasekarshanmugam** [sebastienros/jint#1377] [Q&A] [answered]
I have been looking around sample the samples, tests, but could not find something that helps.

Was looking for a way to create an expression/statement that consumes few input variables, then evaluate these same expression using an input that is a JSON. For e.g 

Expression:
    invoice.id

Input JSON: 
{
    invoice: 
    {
       id: "inv123"
    }
}

Output Expected:
"inv123"

**Comment by lahma** [comment, marked as answer]
```c#
var engine = new Engine();
const string Json = """{ "invoice": { "id": "inv123" } }""";
engine.SetValue("jsonString", Json);
var id = engine.Evaluate("const obj = JSON.parse(jsonString); obj.invoice.id;").AsString();
Assert.Equal("inv123", id);
```

**Reply by rajasekarshanmugam** [reply]
Thank you. Would the below be recommended and performant if we cache the engine instances and execute? Is Invoke threadsafe as no other instance/state data is accessed?

```csharp
using Jint;

var INVOICEJSON = """{"invoice":{"id": "inv123"}}""";
var EXPRESSION2 = """function exec(data){with (JSON.parse(data)){ return invoice.id;}}""";

var engine = new Engine().Execute(EXPRESSION2);
Console.WriteLine(engine.Invoke("exec", INVOICEJSON));
```

**Reply by lahma** [reply]
Hard to say what's the most performant solution as there's little context here. Engine is not thread safe, you cannot access it from multiple threads.

**Reply by rajasekarshanmugam** [reply]
Thank you for the note. To give a brief context, there will be 100s of such simple expressions getting evaluated and some of the expressions will be called too many times from a .NET core web API. Instead of creating an engine and evaluating each expression - which will involve interpreting/executing, I was wondering if we could cache that engine once and evaluate it from many threads. Since each method is not accessing any state (Explicitly) was wondering if that was multi-thread safe. 

if not multi-thread safe, then we have to lock the engine while it's evaluating, it could still be faster than creating an engine each time. Was wondering what the recommendation could be and what folks are doing for scenarios similar to the above.

**Reply by lahma** [reply]
One option is to load outside of Jint as something like Newtonsoft's JObject or the like and just the value into engine, Jint will wrap and use reflection against it. There are a lot of interop test cases in this repo that you can study, also some in readme I belive.
